robot/web/web.go

```diff
@@ -365,42 +364,44 @@ func (svc *webService) updateResources(resources map[resource.Name]resource.Reso
 		groupedResources[n.API] = r
 	}
 
-	apiRegs := resource.RegisteredAPIs()
-	for a, v := range groupedResources {
-		apiResColl, ok := svc.services[a]
-		// TODO(RSDK-144): register new service if it doesn't currently exist
+	// All known APIs have pre-registered resource collections, so loop through them. We want to empty out any collections if
+	// there are no longer resources available in that API.
+	for api, coll := range svc.services {
+		group, ok := groupedResources[api]
 		if !ok {
-			reg, ok := apiRegs[a]
-			var apiResColl resource.APIResourceCollection[resource.Resource]
-			if ok {
-				apiResColl = reg.MakeEmptyCollection()
-			} else {
-				// Log a warning here to remind users to register their APIs. Do not warn if the resource is internal to the RDK or
-				// the resource is handled by a remote with a possibly separate API registration. Modular resources will
-				// have API registrations already and should not reach this point in the method.
-				if a.Type.Namespace != resource.APINamespaceRDKInternal {
-					for n := range v {
-						if !n.ContainsRemoteNames() {
-							svc.logger.Warnw(
-								"missing registration for api, resources with this API will be unreachable through a client", "api", n.API)
-							break
-						}
-					}
+			// create an empty map of resources if one does not exist
+			group = make(map[resource.Name]resource.Resource)
+		}
+		if err := coll.ReplaceAll(group); err != nil {
```

Comment:
I assume this is a similar case where `ReplaceAll` is a directive to remove all old resources and only contain the input `group` resources?

I also assume collections do their own locking (as opposed to this function being wrapped up in a bigger "resource manager" lock)?

I assume the resource manager gets updated with new resources first, then we follow up with updating the web service. But I was curious if we could run into a case where:
* Reconfigure comes in that adds a new wheeled base + adds two new motors.
* The wheeled base is created + added to the resource API collections
* MoveStraightRequest comes in. We service it because it's in the new map
* The wheeled base does its lookup for its left/right motors but fails because they aren't created yet.

I'm pretty certain the above can't happen -- the wheeled base being constructed implies its dependencies are constructed. And we access dependencies through the robot/resource manager.

But I think remotes are interesting (warning: even if this is "a bug" I'm not convinced it's particularly worrying). Consider a scenario where a main part has a wheeled base and a remote part has a motor:
* The remote configures its motor. The motor is constructed and put into the resource graph.
  * The motor can be returned by ResourceNames
  * The antagonist adds a delay before the motor api collection is updated
* The main part reconfigures. It updates its remote resources, sees the motor
* The main part creates/configures the wheeled base. The wheeled base is added to the main parts api collection resource
* A user calls MoveStraight on the main parts wheeled base
* The underlying request to the remote fails because the motor isn't registered in its api collection resource
  * More interesting, the request perhaps only partially fails? For example, the "left motor"s `GoFor` succeeds, but the "right motor" can fail.

Reply:
> I assume this is a similar case where ReplaceAll is a directive to remove all old resources and only contain the input group resources?

yep

`I also assume collections do their own locking (as opposed to this function being wrapped up in a bigger "resource manager" lock)?`

yep

Ya, the first case you described shouldn't happen, but the second case you described can happen. Definitely if we keep the server collections and resource graph in better sync this wouldn't be as much of an issue

Reply:
> Ya, the first case you described shouldn't happen, but the second case you described can happen.

Thanks for confirming!

> Definitely if we keep the server collections and resource graph in better sync this wouldn't be as much of an issue

Agreed. What you have is definitely the right fix for today. But I have general hand-wavy concerns that there's additional subtle undefined behavior due to the discrepancy. I think aligning the two would be a lot of code movement, but largely mechanical? So maybe if someone was really bored one day that could give it a shot to see if there's any unknowns?

Reply:
made a ticket https://viam.atlassian.net/browse/RSDK-8876